### PR TITLE
JW-80 vr_packet manipulation callbacks

### PR DESCRIPTION
**Pull request by sodar**
Depends on https://github.com/codilime/contrail-vrouter/pull/19.

- [x] Testing of vr_pcopy callback

### CODE REVIEW

File: windows/vr_host.c [hunk 1]
```diff
@@ -307,30 +320,94 @@ static int
...
+    ULONG current_offset = offset;
+    while (mdl_data_offset + current_offset >= MmGetMdlByteCount(current_mdl)) {
+        current_offset -= MmGetMdlByteCount(current_mdl) - mdl_data_offset;
+        mdl_data_offset = 0;
+        current_mdl = current_mdl->Next;
+    }
+    mdl_data_offset = current_offset;
+
+    unsigned int copied_bytes = 0;
+    ULONG bytes_to_copy = MmGetMdlByteCount(current_mdl) - mdl_data_offset;
```
Comment (Attil) [inline windows/vr_host.c]: Does `MmGetMdlByteCount` include offset bytes? If not, what would happend in case of offset being greater than byte count?

File: windows/vr_host.c [hunk 2]
```diff
@@ -307,30 +320,94 @@ static int
...
+        return -EFAULT;
+    }
+
+    // Walk through MDL list, until `offset` is reached.
+    PMDL current_mdl = NET_BUFFER_CURRENT_MDL(nb);
+    ULONG mdl_data_offset = NET_BUFFER_CURRENT_MDL_OFFSET(nb);
+    ULONG current_offset = offset;
+    while (mdl_data_offset + current_offset >= MmGetMdlByteCount(current_mdl)) {
+        current_offset -= MmGetMdlByteCount(current_mdl) - mdl_data_offset;
+        mdl_data_offset = 0;
```
Comment (michal-clapinski) [inline windows/vr_host.c]: ULONG current_offset = offset + NET_BUFFER_CURRENT_MDL_OFFSET(nb);
while (current_offset >= MmGetMdlByteCount(current_mdl)) {
    current_offset -= MmGetMdlByteCount(current_mdl);

Wouldn't that snippet do the same thing?

File: windows/vr_host.c [hunk 3]
```diff
@@ -307,30 +320,94 @@ static int
...
+
+    unsigned int copied_bytes = 0;
+    ULONG bytes_to_copy = MmGetMdlByteCount(current_mdl) - mdl_data_offset;
+    unsigned char *mdl_data =
+        (unsigned char *)MmGetSystemAddressForMdlSafe(current_mdl, LowPagePriority | MdlMappingNoExecute);
+    if (!mdl_data) {
+        return -EFAULT;
+    }
+
+    NdisMoveMemory(dst, mdl_data + mdl_data_offset, bytes_to_copy);
```
Comment (michal-clapinski) [inline windows/vr_host.c]: Why do you move data from mdl_data + mdl_data_offset? Could you put a comment here explaining this?

File: windows/vr_host.c [hunk 4]
```diff
@@ -307,30 +320,94 @@ static int
...
+    unsigned short frag_len = (unsigned short)NET_BUFFER_DATA_LENGTH(nb);
+    return frag_len;
 }
 
 static unsigned short
 win_phead_len(struct vr_packet *pkt)
 {
     UNREFERENCED_PARAMETER(pkt);
 
-    /* Dummy implementation */
```
Comment (michal-clapinski) [inline windows/vr_host.c]: Is this correct? You just removed "/* Dummy implementation */" and didn't write any code.

File: windows/vr_host.c [hunk 5]
```diff
@@ -339,8 +416,9 @@ win_pset_data(struct vr_packet *pkt, unsigned short offset)
 {
     UNREFERENCED_PARAMETER(pkt);
     UNREFERENCED_PARAMETER(offset);
+    
+    /* On Windows it is a noop, because there is no `sk_buff->data` pointer equivalent in NET_BUFFER. */
 
-    /* Dummy implementation */
     return;
```
Comment (michal-clapinski) [inline windows/vr_host.c]: Are you sure you want to use useless returns in void functions?

File: windows/vrouter_mod.c [hunk 6]
```diff
@@ -848,5 +849,5 @@ SxExtStartCompleteNetBufferListsIngress(
 	}
 	struct vr_packet* pkt = NetBufferLists->MiniportReserved[0];
 	if (pkt != NULL)
-		windows_host.hos_pfree(pkt, 0);
+		vr_pfree(pkt, 0);
```
Comment (michal-clapinski) [inline windows/vrouter_mod.c]: This change seems dangerous. Filling "vrouter_host" with correct function pointers is done in vrouter.c which isn't included in our project yet (reminder: #define vr_pfree    vrouter_host->hos_pfree).

**Comment by michal-clapinski** [comment]
I think you write to few comments explaining what your code does. This packet/nbl list or whatever is quite hard to understand, so please explain it better.

File: windows/vr_host.c (hunk 3 above)
Comment (Attil) [inline windows/vr_host.c]: MoveMemory actually copies memory.

File: windows/vr_host.c (hunk 2 above)
Comment (sodar) [inline windows/vr_host.c]: I think that in cases when NET_BUFFER_CURRENT_MDL_OFFSET(nb) is non-zero, then the first subtraction from current_offset might give a wrong value.

I am trying now to rewrite this part of the code to make it more clear what I am trying to achieve here.

File: windows/vr_host.c (hunk 4 above)
Comment (sodar) [inline windows/vr_host.c]: Yes.

Briefly - on Linux there is a concept of the head and the fragment of packets. Head is stored alongside with sk_buff structure (sk_buff = packet). Fragment comes into play, when whole packet does not fit in sk_buff - fragments are additionally allocated buffers which can be paged-out.

`win_phead_len` should return length of the head part, but on Windows we only deal with fragments (every of the packet in Windows might be paged out).

File: windows/vr_host.c (hunk 5 above)
Comment (sodar) [inline windows/vr_host.c]: I'll remove it.

File: windows/vr_host.c [hunk 7]
```diff
@@ -139,7 +146,8 @@ win_page_alloc(unsigned int size)
 static void
 win_free(void *mem, unsigned int object)
 {
-    UNREFERENCED_PARAMETER(object);
+	UNREFERENCED_PARAMETER(object);
+
```
Comment (michal-clapinski) [inline windows/vr_host.c]: correct tabs

File: windows/vr_host.c [hunk 8]
```diff
@@ -234,7 +242,7 @@ win_palloc(unsigned int size)
     return win_get_packet(nbl, NULL);
 }
 
-static void
+void
```
Comment (michal-clapinski) [inline windows/vr_host.c]: how do you determine where to put static and where to not?

File: include/vr_windows.h [hunk 9]
```diff
@@ -40,6 +40,8 @@ int vr_hash_nic(NDIS_IF_COUNTED_STRING string);
 struct vr_assoc* vr_get_assoc(NDIS_IF_COUNTED_STRING string);
 void vr_set_assoc_oid(NDIS_IF_COUNTED_STRING name, NDIS_SWITCH_PORT_ID port_id);
 
+struct host_os * vrouter_get_host(void);
```
Comment (michal-clapinski) [inline include/vr_windows.h]: why is that here and also in windows/vr_host.c (line 13)?

File: windows/vr_host.c [hunk 10]
```diff
@@ -13,6 +14,8 @@ struct host_os * vrouter_get_host(void);
 
 #define vrouter_host (vrouter_get_host())
```
Comment (michal-clapinski) [inline windows/vr_host.c]: I know it's not part of this pull request, but why is it here?

File: windows/vr_host.c [hunk 11]
```diff
@@ -325,9 +333,21 @@ win_pexpand_head(struct vr_packet *pkt, unsigned int hspace)
...
+        return;
+
+    PNET_BUFFER nb = NET_BUFFER_LIST_FIRST_NB(nbl);
+    if (!nb)
+        return;
+
+    PMDL current_mdl = NET_BUFFER_CURRENT_MDL(nb);
+    pkt->vp_head =
+        (unsigned char*)MmGetSystemAddressForMdlSafe(current_mdl, LowPagePriority | MdlMappingNoExecute);
+    pkt->vp_data = 0;
```
Comment (michal-clapinski) [inline windows/vr_host.c]: I think NULL is a better idea than 0. Also, why in win_preset do you set vp_data to 0 but not vp_head?

File: windows/vr_host.c [hunk 12]
```diff
@@ -347,30 +367,126 @@ static int
...
+
+    /* Check if requested data offset lies in the NET_BUFFER's current MDL. */
+    PMDL current_mdl = NET_BUFFER_CURRENT_MDL(nb);
+    if (NET_BUFFER_CURRENT_MDL_OFFSET(nb) + offset >= MmGetMdlByteCount(current_mdl)) {
+        /* Requested offset lies outside of the first MDL => traverse MDL list until offset is reached */
+        offset -= MmGetMdlByteCount(current_mdl) - NET_BUFFER_CURRENT_MDL_OFFSET(nb);
+        current_mdl = current_mdl->Next;
+        if (!current_mdl) {
+            return -EFAULT;
+        }
```
Comment (michal-clapinski) [inline windows/vr_host.c]: I think that if "data_offset + (ULONG)offset + (ULONG)len > data_size" wasn't true than this will never happen, am I right?

File: windows/vr_host.c [hunk 13]
```diff
@@ -347,30 +367,126 @@ static int
...
+        while (offset >= MmGetMdlByteCount(current_mdl)) {
+            offset -= MmGetMdlByteCount(current_mdl);
+            current_mdl = current_mdl->Next;
+            if (!current_mdl) {
+                return -EFAULT;
+            }
```
Comment (michal-clapinski) [inline windows/vr_host.c]: Comment above also true for this.

File: windows/vr_host.c [hunk 14]
```diff
@@ -347,30 +367,126 @@ static int
...
+    ULONG copied_bytes = 0;
+    ULONG bytes_left_in_first_mdl = MmGetMdlByteCount(current_mdl) - offset;
+    if (bytes_left_in_first_mdl <= len) {
+        NdisMoveMemory(dst, mdl_data + offset, bytes_left_in_first_mdl);
+        copied_bytes += bytes_left_in_first_mdl;
+    } else {
+        /* All of the requested data lies in `current_mdl` */
+        NdisMoveMemory(dst, mdl_data + offset, len);
+        copied_bytes += len;
+    }
```
Comment (michal-clapinski) [inline windows/vr_host.c]: Why is this code outside of the loop below? Could you merge them?

File: windows/vr_host.c [hunk 15]
```diff
@@ -347,30 +367,126 @@ static int
...
+        } else {
+            /* Otherwise copy only the necessary amount. */
+            NdisMoveMemory(dst + copied_bytes, mdl_data, left_to_copy);
+            copied_bytes += left_to_copy;
+        }
+
+        current_mdl = current_mdl->Next;
+    }
+
+    if (copied_bytes < len) {
```
Comment (michal-clapinski) [inline windows/vr_host.c]: As I said before, I think you already checked it by "data_offset + (ULONG)offset + (ULONG)len > data_size". Do you think multiple checks are necessary?

File: windows/vrouter_mod.c [hunk 16]
```diff
@@ -23,7 +23,10 @@ NDIS_HANDLE SxNBLPool = NULL;
 */
 PNDIS_RW_LOCK_EX AsyncWorkRWLock = NULL;
 
-extern struct host_os windows_host; // TODO: Delete after vr_pfree works
+extern struct host_os windows_host;
```
Comment (michal-clapinski) [inline windows/vrouter_mod.c]: Looks like you were to delete this. In "include/vrouter.h" you have "extern struct host_os *vrouter_host;". Maybe use this pointer instead of writing another extern?

File: windows/vrouter_mod.c [hunk 17]
```diff
@@ -23,7 +23,10 @@ NDIS_HANDLE SxNBLPool = NULL;
...
+
+/* Defined in windows/vr_host.c */
+extern void win_pfree(struct vr_packet *pkt, unsigned short reason);
```
Comment (michal-clapinski) [inline windows/vrouter_mod.c]: Shouldn't this be in include/vr_windows.h or in some other header?

File: windows/vrouter_mod.c [hunk 18]
```diff
@@ -130,6 +133,8 @@ SxExtCreateSwitch(
 		return NDIS_STATUS_RESOURCES;
 	}
 
+	SxNBLPool = vrouter_generate_pool();
```
Comment (michal-clapinski) [inline windows/vrouter_mod.c]: Why did you make this change?

File: windows/vr_host.c (hunk 8 above)
Comment (sodar) [inline windows/vr_host.c]: This `win_pfree` function is used also in `windows/vrouter_mod.c` so it cannot be static.

File: windows/vr_host.c (hunk 11 above)
Comment (sodar) [inline windows/vr_host.c]: `vp_data`, `vp_tail` and `vp_end` are stored as offsets in dp-core and this is why I set it to zero (packet data begins at `vp_head + vp_data`).

File: windows/vr_host.c (hunk 12 above)
Comment (sodar) [inline windows/vr_host.c]: Theoretically yes - if this condition is false, then required MDLs exist.

But if Next pointers will be NULL (for example - our programming error or, less likely, operating system error), then functions which take pointer to MDL as parameter will result in blue screen.

Comment (sodar) [inline windows/vr_host.c]: Maybe better option is to log a critical error, return the error code.

File: windows/vr_host.c (hunk 14 above)
Comment (sodar) [inline windows/vr_host.c]: In the loop below copies are done from `mdl_data` without any offset. This one is the special case when we copy from the first MDL where requested data occurs and we need to take an offset into account.

File: windows/vr_host.c (hunk 15 above)
Comment (sodar) [inline windows/vr_host.c]: Same reasoning as above - we could reach NULL pointer on MDL list before `copied_bytes` will reach `len`.

File: include/vr_windows.h (hunk 9 above)
Comment (sodar) [inline include/vr_windows.h]: It should be removed from windows/vr_host.c.

File: windows/vrouter_mod.c (hunk 16 above)
Comment (sodar) [inline windows/vrouter_mod.c]: `windows_host` struct is not used in `vrouter_mod.c`, so this line can be removed.

File: windows/vrouter_mod.c (hunk 18 above)
Comment (sodar) [inline windows/vrouter_mod.c]: It is my error. I've done it when I merged @Attil recent changes.

File: windows/vrouter_mod.c (hunk 17 above)
Comment (sodar) [inline windows/vrouter_mod.c]: Yes, it should be. I'll add it to include/vr_windows.h.

File: windows/vr_host.c (hunk 10 above)
Comment (sodar) [inline windows/vr_host.c]: Actually it is not needed. I'll remove it.

File: windows/vr_host.c (hunk 8 above)
Comment (sodar) [inline windows/vr_host.c]: I'll change it to static and I'll call this function through `host_os`. This will make our version consistent with other implementations.
